diff --git a/src/compiler/scala/tools/nsc/Global.scala b/src/compiler/scala/tools/nsc/Global.scala
index b0288c0..f6cc1d2 100644
--- a/src/compiler/scala/tools/nsc/Global.scala
+++ b/src/compiler/scala/tools/nsc/Global.scala
@@ -201,6 +201,9 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
   val treeBrowser = treeBrowsers.create()
 
   // ------------ Hooks for interactive mode-------------------------
+  
+  /** Tracks errors related to qualifiers */
+  var qualToRecoveredType = new mutable.HashMap[Tree, Type]
 
   /** Called every time an AST node is successfully typechecked in typerPhase.
    */
diff --git a/src/compiler/scala/tools/nsc/interactive/Global.scala b/src/compiler/scala/tools/nsc/interactive/Global.scala
index 56617d0..fc40218 100644
--- a/src/compiler/scala/tools/nsc/interactive/Global.scala
+++ b/src/compiler/scala/tools/nsc/interactive/Global.scala
@@ -847,9 +847,10 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
       (!implicitlyAdded || m.implicitlyAdded)
 
     def add(sym: Symbol, pre: Type, implicitlyAdded: Boolean)(toMember: (Symbol, Type) => M) {
-      if ((sym.isGetter || sym.isSetter) && sym.accessed != NoSymbol) {
-        add(sym.accessed, pre, implicitlyAdded)(toMember)
-      } else if (!sym.name.decodedName.containsName(Dollar) && !sym.isSynthetic && sym.hasRawInfo) {
+      //if ((sym.isGetter || sym.isSetter) && sym.accessed != NoSymbol) {
+      //  add(sym.accessed, pre, implicitlyAdded)(toMember)
+      //} else if (!sym.name.decodedName.containsName(Dollar) && !sym.isSynthetic && sym.hasRawInfo) {
+      if (!sym.name.decodedName.containsName(Dollar) && !sym.isSynthetic && sym.hasRawInfo) {
         val symtpe = pre.memberType(sym) onTypeError ErrorType
         matching(sym, symtpe, this(sym.name)) match {
           case Some(m) =>
@@ -969,7 +970,7 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
     }
 
     //print("add members")
-    for (sym <- ownerTpe.members)
+    for (sym <- ownerTpe.resultType.members)
       addTypeMember(sym, pre, sym.owner != ownerTpe.typeSymbol, NoSymbol)
     members.allMembers #:: {
       //print("\nadd pimped")
diff --git a/src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala b/src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala
index 9e9b8b9..a1cfa17 100644
--- a/src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala
@@ -297,6 +297,8 @@ trait ContextErrors {
 
       //typedSelect
       def NotAMemberError(sel: Tree, qual: Tree, name: Name) = {
+		global.qualToRecoveredType += (sel  -> qual.tpe.widen)
+		global.qualToRecoveredType += (qual -> qual.tpe.widen)
         def errMsg = {
           val owner            = qual.tpe.typeSymbol
           val target           = qual.tpe.widen
